### PR TITLE
CLI generator script

### LINKED ISSUE #17: Stub out generator script
We eventually want to have a generator script to allow folks to answer a few different questions and generate a scaffolded app.

This seems to be mostly including a script in the ```bin``` section of the package.json which runs the process to scaffold out the project.

### PR DESCRIPTION
**Pull request by penelopeysm**
Closes #17 

## Test this out!!

```
npm create @uatp/web@latest
```

Select Rust, Python, or none. Then `cd` to wherever you set it up, and:

```
cd template_app
pnpm i
pnpm rust    # if you used Rust
pnpm python  # if you used Python
pnpm dev
```

and view on http://localhost:5173.

### Readme

I've updated quite a few parts of the readme, so maybe it's a good starting point of what the repo looks like now.

https://github.com/Urban-Analytics-Technology-Platform/web-app-template/blob/create-template/README.md

### `@uatp` on npm

I made a `@uatp` organisation on npm. (I mean, why not? But also to get the command above to work, the package has to be on the npm registry.) I will post about this on Slack and add everyone who wants to be added.

### How to review this PR?!

Obviously, the code changes associated with this PR are slightly ridiculous and virtually impossible to review. I would like to suggest that you not evaluate it based on the code, but rather the effect ('try this out' section above).

Roughly, the steps I've had to do are:

 - Reorganise the repository. I modelled this after SvelteKit's example: https://github.com/sveltejs/kit/tree/main.

 - Write a script that runs an interactive prompt, copies the template package to wherever the user wants, and **modifies the code** inside it to include/remove languages. This script is itself part of a package, called `@uatp/create-web` (essentially, when you run `npm create @uatp/web`, npm searches for this package and runs its main script).

 - I renamed the three packages respectively to all be `@uatp/...`; the library is now `components`, the template is `template`, and the web creation script is `create-web`.

 - I published two of the packages to npm so that they could be tested. One is the `create-web` one, of course: https://www.npmjs.com/package/@uatp/create-web
   The other is `components`. Why does components need to be published at this stage? See below.

As part of this PR, I've also removed some of Dustin's backend-loading code; I've replaced it with something I understood a bit better. Sorry!! For example the loading screen I think no longer appears and the initialisation calls are rather more dumbed down. My hope is that it is not too hard to add these back in as part of a separate PR.

## Everything below this...

... is just finer detail. You don't really have to read it unless it interests you, or if you want to review this PR in more depth than 'works as expected'.

### What does the repository reorganisation imply?

This PR will change how we view this repository. The repository _contains_ the template, but the repository is not _the_ template itself; it is code which generates the template.

Thus, most files will have to be considered from both a developer and a user standpoint. Take the simple example of a license. If we add a MIT license to the top level directory, this indicates that the entire repo structure is MIT-licensed. It does not mean that when someone runs `npm create @uatp/web` they will get a MIT license in their project: that is determined by the license file we put inside the template. This logic extends to other things, like documentation, as well as `package.json`.

From a workflow perspective, what this means is that you would only git clone this repository if you were a _developer_ of the template. If you were just using the template, you'd just run `npm create @uatp/web` and be done.

I've documented some of these in the top-level repo README. I'm sure it can be more detailed, though.

### The `create-web` script modifies the code?

Yes, that was _really_ fun. You can see the script in [`packages/create-web/bin.js`](https://github.com/Urban-Analytics-Technology-Platform/web-app-template/blob/create-template/packages/create-web/bin.js) and [`packages/create-web/index.js`](https://github.com/Urban-Analytics-Technology-Platform/web-app-template/blob/create-template/packages/create-web/index.js).

This was a seriously difficult challenge because we want to accomplish several things:
1. **Retain the functionality of the template itself**, so that local development on the template isn't prohibitively difficult. If the template cannot be built as a package itself and run with `pnpm dev`, it's going to be very hard to iterate on it.
2. **Avoid excessive code duplication**. For example, we could simply have three different template repos, one for Python, one for Rust, and one with neither. The script would then just copy the appropriate directory. While this would work, it's not good as changes to Svelte code (for example) have to be made three times.
3. **A middle ground between generality and complexity**. We want to be able to work on the template and somehow use comments to indicate "_this bit of code_ needs to go into the Rust part". We don't want to write a script that excises specific line numbers in specific files because that will break (not general enough). On the other hand, I'm also not keen on reimplementing a whole templating language like Jinja (too general; and besides, Jinja code isn't valid HTML, see (1)).

I've written up notes on how I did this, and added them as documentation accompanying this PR: https://github.com/Urban-Analytics-Technology-Platform/web-app-template/blob/create-template/packages/create-web/DEVELOPER_NOTES.md

### Did you write tests?

Yes, in the [`ci.yml` GitHub Action workflow](https://github.com/Urban-Analytics-Technology-Platform/web-app-template/blob/create-template/.github/workflows/ci.yml) I've added tests which generate new web apps with Rust/Python/no backend and test that they build correctly.

### Why didn't you use TypeScript for `create-web`?

Because then `npm create @uatp/web` has to needlessly pull in more dependencies before running, and we'd also have to add more bloat like tsconfig, eslint, etc.. For a script which is quite small and honestly isn't too hard to follow, I thought this was overkill. (As you can see in the commit messages, I started with TS and then ditched it.)

### Why publish `@uatp/components` now to npm?

In this monorepo structure, our template app depends on the local version of components. This is specified using `workspace: *` in its `package.json`. This is exactly what we want for testing because we always want to test our local version of the template against the local version of the components.

When somebody sets up a new app using this template, they aren't going to have a workspace complete with the source code of the `components` library. (I mean, they _could_, in the sense that we could make the `create-web` script set that up for them — but it would be terrible practice because it would mean that every UA web app would bundle its own copy of the library.)

Thus, when setting up a new app for them, the `create-web` script has to modify `package.json` and specify an actual version of `components` to depend on. In general, we should assume that each version of the template app has a specific, known, version of `components` which it can reliably be built against. It will then be the user's responsibility to upgrade the version dependency if they want newer updates. (Note that this is true of every dependency in the template; `components` is not special in this regard.)

So, when a new user runs `npm create @uatp/web`, they will get a `package.json` that points to some version of `@uatp/components`, and if we don't publish components to npm now, they'll get an error when they try to `pnpm i`.

### Does `@uatp/template` need to be published as well?

I don't think so, no.

### But `@uatp/create-web` is published?

Yes, it's published so that `npm create @uatp/web` works correctly.

### CODE REVIEW

**Comment by penelopeysm** [comment]
Some problems I've noticed so far:

- we are currently specifying that the template_app package installs its uatp-components as a library from the current workspace.

   https://github.com/Urban-Analytics-Technology-Platform/web-app-template/blob/5a189ed3d0a6b22fdaa45dad3c8b3fde440a144d/packages/create-ua-web/template/template_app/package.json#L51

   This is not good because we should not be making users include the entire uatp-components library in their app (and then publish it to github!!). uatp-components should be published as a package on the npm registry and then the template_app package should depend on it.

**Comment by penelopeysm** [comment]
Another potentially controversial thought: The _app template itself_ should not be tested. Instead we should use the create-ua-web script to generate new apps that can be tested.

The reason for this is because it's too hard to have the _template repo_ also be valid code.

For example, we could do something like this and have the create-ua-web script remove any lines that don't apply by checking the line-ending comments. However, this will fail to compile because it is imported twice.

```typescript
  import workerWrapper from "./rust_worker?worker";  // IF RUST
  import { type Backend } from "./rust_worker";  // IF RUST
  import workerWrapper from "./python_worker?worker";  // IF PYTHON
  import { type Backend } from "./python_worker";  // IF PYTHON
  */
```

On the other hand, this code will compile, but it cannot be tested:

```typescript
  /* IF RUST
  import workerWrapper from "./rust_worker?worker";
  import { type Backend } from "./rust_worker";
  */
  /* IF PYTHON
  import workerWrapper from "./python_worker?worker";
  import { type Backend } from "./python_worker";
  */
```

File: packages/create-web/index.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    // Remove all empty commands and any entries that still start with @@
+    for (const [key, val] of scripts.entries()) {
+        if (val === "" || key.startsWith("@@")) {
+            scripts.delete(key);
+        }
+    }
+    packageJson.scripts = Object.fromEntries(scripts);
+
+    // Fix dependencies. I don't know any way to not hardcode this
+    packageJson.dependencies["@uatp/components"] = "0.0.1";
```
Comment (penelopeysm) [inline packages/create-web/index.js:173]: Obtain latest version of @uatp/components from npm https://www.npmjs.com/package/latest-version

Comment (penelopeysm) [reply packages/create-web/index.js:173]: No, don't do that, each version of create-web needs to be pinned to one version of components

Comment (penelopeysm) [reply packages/create-web/index.js:173]: Create issue about this & "automatic updating" or to be precise informing the user when they can update

File: packages/create-web/index.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    backendLanguage, // "python" | "rust" | "none"
+) {
+    const pathToTemplate = path.join(
+        path.dirname(fileURLToPath(import.meta.url)),
+        "template"
+    );
+
+    // TODO: Don't copy things that match gitignore
+    // Copy files that will always be there
+    cp_r(".gitignore", pathToTemplate, dstDir);
```
Comment (penelopeysm) [inline packages/create-web/index.js:17]: gitignore file is not present on npm package

File: packages/create-web/index.js (hunk 1 above)
Comment (griff-rees) [reply packages/create-web/index.js:173]: See https://github.com/Urban-Analytics-Technology-Platform/web-app-template/issues/29

**Comment by penelopeysm** [comment]
Merging after discussion on 6 Jul!

**Review by dabreegster — COMMENTED** [review]
Super awesome!!

> As part of this PR, I've also removed some of Dustin's backend-loading code; I've replaced it with something I understood a bit better. Sorry!! For example the loading screen I think no longer appears and the initialisation calls are rather more dumbed down. My hope is that it is not too hard to add these back in as part of a separate PR.

That's fine, the old stuff was probably overly complicated. The UX and DX of the loading screens with messages aren't great yet, so I'd also be fine leaving them out of the template for now. They won't be hard to incorporate later

File: packages/create-web/index.js (hunk 2 above)
Comment (penelopeysm) [reply packages/create-web/index.js]: Resolved, see code in later commits, and also https://stackoverflow.com/a/78712387/26116613 for explanation

File: packages/create-web/template/rust_backend/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        // serde is used to parse a JSON object into a Rust struct
+        let gj: FeatureCollection = serde_wasm_bindgen::from_value(input)?;
+        let mut new_gj = gj.clone();
+
+        // Perform your calculations here. You can extract this block into a separate crate too
+        let new_features: Vec<Feature> = gj
+            .features
+            .into_iter()
+            .map(|mut feat| {
+                let mut properties = match feat.properties {
```
[the remark this answers is not in the document]
Comment (penelopeysm) [reply packages/create-web/template/rust_backend/src/lib.rs:57]: Edited in #31!
